index.js

```diff
@@ -1,22 +1,25 @@
 
-module.exports = function(router) {
-  return function(paramName, handler) {
-    return router.param(paramName, function(req, res, next, val, name) {
-      if (handler instanceof Function) {
-        var newVal = handler(val)
-        if (!newVal) {
-          return next('route'); // bypass current route
-        }
-        req.params[name] = newVal;
-      } else if (handler instanceof RegExp) {
-        if (!handler.test(val)) {
-          return next('route'); // bypass current route
-        }
-        req.params[name] = val;
-      } else {
-        throw new Error('Unsupported param handler. Either function or Regexp expected');
-      }
-      return next();
-    });
-  };
-};
+module.exports = function(handler) {
+  if (!(handler instanceof Function || handler instanceof RegExp))
+    throw new Error('Unsupported param handler. Either function or Regexp expected')
+
+  return function(req, res, next, val, name) {
+
```

Comment:
We need to use value of `req.params[name]`, because `req.params[name]` can be modified early but express does not change `val` for us (`val` always has same original value):
```js
val = req.params[name];
```